src/providers/tpm_provider/mod.rs

```diff
@@ -0,0 +1,462 @@
+        }
+    }
+
+    fn create_key(&self, app_name: ApplicationName, op: OpCreateKey) -> Result<ResultCreateKey> {
+        if op.key_attributes.key_type != KeyType::RsaKeypair
+            || op.key_attributes.algorithm != Algorithm::sign(SignAlgorithm::RsaPkcs1v15Sign, None)
+        {
+            error!("The TPM provider currently only supports creating RSA key pairs for signing and verifying.");
+            return Err(ResponseStatus::UnsupportedOperation);
+        }
+
+        let key_name = op.key_name;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+        // This should never panic on 32 bits or more machines.
+        let key_size = std::convert::TryFrom::try_from(op.key_attributes.key_size)
+            .expect("Conversion to usize failed.");
+
+        let mut store_handle = self.key_id_store.write().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let (key_context, auth_value) = esapi_context
+            .create_rsa_signing_key(key_size, AUTH_VAL_LEN)
+            .or_else(|e| {
+                error!("Error creating a RSA signing key: {}.", e);
+                dbg!(e);
+                Err(ResponseStatus::PsaErrorHardwareFailure)
+            })?;
+
+        insert_password_context(
+            &mut *store_handle,
+            key_triple,
+            PasswordContext {
+                context: key_context,
+                auth_value,
+            },
+        )?;
+
+        Ok(ResultCreateKey {})
+    }
+
+    fn import_key(&self, app_name: ApplicationName, op: OpImportKey) -> Result<ResultImportKey> {
+        if op.key_attributes.key_type != KeyType::RsaPublicKey
+            || op.key_attributes.algorithm != Algorithm::sign(SignAlgorithm::RsaPkcs1v15Sign, None)
+        {
+            error!(
+                "The TPM provider currently only supports importing RSA public key for verifying."
+            );
+            return Err(ResponseStatus::UnsupportedOperation);
+        }
+
+        let key_name = op.key_name;
+        let key_triple = KeyTriple::new(app_name, ProviderID::TpmProvider, key_name);
+        let key_data = op.key_data;
+
+        let mut store_handle = self.key_id_store.write().expect("Key store lock poisoned");
+        let mut esapi_context = self
+            .esapi_context
+            .lock()
+            .expect("ESAPI Context lock poisoned");
+
+        let public_key: RsaPublicKey = serde_asn1_der::from_bytes(&key_data).or_else(|err| {
+            error!("Could not deserialise key elements: {}.", err);
+            Err(ResponseStatus::PsaErrorCommunicationFailure)
+        })?;
+        if public_key.public_exponent.to_bytes_be().1 != PUBLIC_EXPONENT {
+            error!("The TPM Provider only supports 0x101 as public exponent for RSA public keys, {:?} given.", public_key.public_exponent.to_bytes_be());
```

Comment:
Maybe we should have an issue to change that

Reply:
It is a limitation given by the TSS 2.0 specs, check https://trustedcomputinggroup.org/wp-content/uploads/TPM-Rev-2.0-Part-2-Structures-01.38.pdf section 12.2.3.5 "`TPMS_RSA_PARMS`", page 133

> A TPM compatible with this specification and supporting RSA shall support two primes and an exponentof  zero.  Support  for  other  values  is  optional.  Use  of  other  exponents  in  duplicated  keys  is  not  recommended because the resulting keys would not be interoperable with other TPMs

And an exponent of zero maps to `0x101`:

> When  zero,  indicates  that  the  exponent  is  the  default  of 2^16 + 1

Reply:
Given that the feature is optional, we could introduce it (as an option :D ) - default it to 0, but give ppl the possibility to change it

I'll raise an issue but not set the v1.0 Milestone on it